src/main/core/logger/shadow_logger.rs

```diff
@@ -0,0 +1,426 @@
+    // self.records. If `done_sender` is provided, it's notified after the flush
+    // has completed.
+    fn flush_records(&self, done_sender: Option<Sender<()>>) -> std::io::Result<()> {
+        use std::io::Write;
+
+        // Only flush records that are already in the queue, not ones that
+        // arrive while we're flushing. Otherwise callers who perform a
+        // synchronous flush (whether this flush operation or another one that
+        // arrives while we're flushing) will be left waiting longer than
+        // necessary. Also keeps us from holding the stdout lock indefinitely.
+        let mut toflush = self.records.len();
+
+        let stdout_unlocked = std::io::stdout();
+        let stdout_locked = stdout_unlocked.lock();
+        let mut stdout = std::io::BufWriter::new(stdout_locked);
+        while toflush > 0 {
+            let record = match self.records.pop() {
+                Some(r) => r,
+                None => {
+                    // This can happen if another thread panics while the
+                    // logging thread is flushing. In that case both threads
+                    // will be consuming from the queue.
+                    break;
+                }
+            };
+            toflush -= 1;
+            {
+                let parts = TimeParts::from_nanos(record.wall_time.as_nanos());
+                write!(
+                    stdout,
+                    "{:02}:{:02}:{:02}.{:06}",
+                    parts.hours,
+                    parts.mins,
+                    parts.secs,
+                    parts.nanos / 1000
+                )?;
+            }
+            write!(stdout, " [{}]", record.thread_name)?;
+            if let Some(sim_time) = record.sim_time {
+                let parts = TimeParts::from_nanos(sim_time.as_nanos());
+                write!(
+                    stdout,
+                    " {:02}:{:02}:{:02}.{:09}",
+                    parts.hours, parts.mins, parts.secs, parts.nanos
+                )?;
+            } else {
+                write!(stdout, " n/a")?;
+            }
+            write!(
+                stdout,
+                " [{level}] [{host}] [{file}:{line}] [{module}] {msg}\n",
+                level = record.level,
+                host = record
+                    .host_name
+                    .as_ref()
+                    .map(|s| s.as_str())
+                    .unwrap_or("n/a"),
+                file = record
+                    .file
+                    .map(|f| if let Some(sep_pos) = f.rfind('/') {
+                        &f[(sep_pos + 1)..]
+                    } else {
+                        f
+                    })
+                    .unwrap_or("n/a"),
+                line = record
+                    .line
+                    .map(|l| format!("{}", l))
+                    .as_ref()
+                    .map(|s| s.as_str())
+                    .unwrap_or("n/a"),
```

Comment:
Maybe just print "-1" instead of "n/a" here if not available to avoid the extra string allocation from `format!()`?

Reply:
Broke up this `write!` to avoid the allocation while keepin "n/a" instead of "-1".